palmetto/palm-cli
# [WIP] [FEATURE] Multi service docker support

## Pull request checklist

Before submitting your PR, please review the following checklist:

- [ ] All new and existing tests pass locally (`palm test`)
- [ ] Lint (`palm lint`) has passed locally and any fixes were made for failures

### Breaking changes

- [ ] Check if this pull request introduces a breaking change

## What does this implement/fix? Explain your changes.

Adds support for projects with multiple services in docker-compose.
Palm was originally built for very simple applications with a single 'primary' service. This has worked well for our needs, but doesn't allow folks to use palm for more complex projects.

## Does this close any currently open issues?

#16

## Any other comments?

I also added some new utilities to support this functionality:
- choice prompt: easily create a 'choose from list' interface for any CLI command.
- `environment.get_plugin` allows you to get a plugin by name, useful if a plugin provides a method for re-use
- docker details: info about the docker-compose file is now available via `environment.palm`

Housekeeping:
- Removed `palm scaffold config` - this was replaced by `palm init` long ago and is unused.

**Note:** This is still a WIP. I need to add tests and docs for all of this. Opening the PR now to get feedback on the functionality from users.

## Testing instructions

To test this functionality, you will need a codebase with multiple docker services in your compose file.
Configuration:
1. `palm init` to set your repo up for using palm if you haven't already done so.
2. Add the multi_service plugin to your repo's `.palm/config.yaml` it should look like this:
```yaml
plugins:
  - multi_service
```
3. `palm` to see available commands, you should see a section for 'Multi Service' with up/down/logs/exec - try them out!
4. Try writing a custom palm command to execute something useful in one (or more) of your containers
5. Try to break it!

## Where has this been tested?

**Operating System:** MacOS

**Platform:** Python 3.9
